### DISCUSSION TITLE
Optional function args?

### DISCUSSION THREAD

**Discussion by prescience-data** [colinhacks/zod#401] [Q&A]
Wanted to see if there was a way to implement a function shape that may have zero to one args?

What I've tried:

```typescript
export const hostAttempt1 = z
  .function()
  .args(z.union([z.string(), z.undefined()]))
  .returns(z.string())
  .default(() => "127.0.0.1")

export const hostAttempt2 = z
  .function()
  .args(z.string().optional())
  .returns(z.string())
  .default(() => "127.0.0.1")

export const hostAttempt3 = z
  .function()
  .args(z.union([z.string(), z.undefined()]).optional())
  .returns(z.string())
  .default(() => "127.0.0.1")

```

No matter what I try I seem to get:

```
TS2554: Expected 1 arguments, but got 0.
```

**Comment by colinhacks** [comment]
This isn't currently possible becuase Zod doesn't support truly optional tuple values, e.g. `[string?, number?]`. There's a deeper discussion about this in #149. Now that the minimum TS version for Zod 3 is TS4.1, this is now possible: https://github.com/colinhacks/zod/issues/149#issuecomment-692196245 though I haven't gotten around to it. PR welcome!
